apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+import * as Immutable from 'immutable';
+import Observer from '../../../Observer';
+import {update as updateWatchExpressions} from '../../../redux/watchedExpressions';
+import CommandHistory from './CommandHistory';
+import JSInterpreter from '../../../JSInterpreter';
+import watchedExpressions from '@cdo/apps/redux/watchedExpressions';
+
+const WATCH_TIMER_PERIOD = 250;
+const INITIALIZE = 'jsdebugger/INITIALIZE';
+const ATTACH = 'jsdebugger/ATTACH';
+const DETACH = 'jsdebugger/DETACH';
+const MIRROR_INTERPRETER_STATE = 'jsdebugger/MIRROR_INTERPRETER_STATE';
+const APPEND_LOG = 'jsdebugger/APPEND_LOG';
+const CLEAR_LOG = 'jsdebugger/CLEAR_LOG';
+
+// State model
+const JSDebuggerState = Immutable.Record({
+  jsInterpreter: null,
+  runApp: null,
+  observer: null,
+  watchIntervalId: null,
+  commandHistory: null,
+  logOutput: '',
+  jsInterpreterNextStep: null,
+  jsInterpreterPaused: false,
+});
+
+export function getRoot(state) {
+  return state.jsdebugger;
+}
+
+export function getCommandHistory(state) {
+  return getRoot(state).commandHistory;
+}
+
+export function getJSInterpreter(state) {
+  return getRoot(state).jsInterpreter;
+}
+
+function getRunApp(state) {
+  return getRoot(state).runApp;
+}
+
+export function isPaused(state) {
+  return getRoot(state).jsInterpreterPaused;
+}
+
+function getObserver(state) {
+  return getRoot(state).observer;
+}
+
+function getWatchIntervalId(state) {
+  return getRoot(state).watchIntervalId;
+}
+
+export function isAttached(state) {
+  return !!getJSInterpreter(state);
+}
+
+export function canRunNext(state) {
+  return (
+    isPaused(state) &&
+    getRoot(state).jsInterpreterNextStep === JSInterpreter.StepType.RUN
+  );
+}
+
+export function getLogOutput(state) {
+  return getRoot(state).logOutput;
+}
+
+export const selectors = {
+  getRoot,
+  getCommandHistory,
+  getJSInterpreter,
+  isPaused,
+  isAttached,
+  canRunNext,
+  getLogOutput,
+};
+
+// actions
+export function initialize({runApp}) {
+  return {type: INITIALIZE, runApp};
+}
+
+export function appendLog(output) {
+  return {type: APPEND_LOG, output};
+}
+
+export function clearLog() {
+  return {type: CLEAR_LOG};
+}
+
+function mirrorInterpreterState(jsInterpreter) {
+  return {type: MIRROR_INTERPRETER_STATE, jsInterpreter};
```

Comment:
I really don't want this to be an action. The issue is that the interpreter itself owns some state that the debugger depends on. So we need some way of triggering updates to the debugger UI when the interpreter's state changes. In `JsDebuggerUI`, this is accomplished imperatively, with interpreter events directly triggering DOM modification. Since react/redux is a functional model, we have to reflect the state used to make the previously imperative decisions inside of our redux store. Hence the need to mirror the state from the js interpreter at key moments.